src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -219,27 +225,89 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
         {
-            var diagnostics = _diagnosticService.GetDiagnostics(document.Project.Solution.Workspace, document.Project.Id, document.Id, null, false, cancellationToken);
+            var diagnostics = _diagnosticService.GetDiagnostics(document.Project.Solution.Workspace, document.Project.Id, document.Id, null, false, cancellationToken)
+                                                .Where(IncludeDiagnostic);
+
             var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
 
-            return diagnostics.Select(diagnostic => new LanguageServer.Protocol.Diagnostic
+            // Razor documents can import other razor documents
+            // https://docs.microsoft.com/en-us/aspnet/core/mvc/views/layout?view=aspnetcore-3.1#importing-shared-directives
+            // https://docs.microsoft.com/en-us/aspnet/core/blazor/layouts?view=aspnetcore-3.1#centralized-layout-selection
+            // The imported files contents are added to the content of the generated C# file, so we get diagnostics
+            // for both the c# contents in the original razor document and for any of the content in any of the imported files when we query diagnostics for the generated C# file.
+            // These diagnostics will be reported with DiagnosticDataLocation.OriginalFilePath = generated C# file, and DiagnosticDataLocation.MappedFilePath = imported razor file.
+            // This means that in general we could have diagnostics produced by one generated file that map to many different actual razor files.
+            // We can't filter them out as we don't know which razor file(s) the underlying generated C# document actually maps to, and which are just imported.
+            // So we publish them all and let them get de-duped.
+            var fileUriToDiagnostics = diagnostics.GroupBy(diagnostic => GetDiagnosticUri(document, diagnostic)).ToDictionary(
+                group => group.Key,
+                group => group.Select(diagnostic => ConvertToLspDiagnostic(diagnostic, text)).ToArray());
+            return fileUriToDiagnostics;
+
+            static Uri GetDiagnosticUri(Document document, DiagnosticData diagnosticData)
             {
-                Code = diagnostic.Id,
-                Message = diagnostic.Message,
-                Severity = ProtocolConversions.DiagnosticSeverityToLspDiagnositcSeverity(diagnostic.Severity),
-                Range = ProtocolConversions.TextSpanToRange(DiagnosticData.GetExistingOrCalculatedTextSpan(diagnostic.DataLocation, text), text),
-                // Only the unnecessary diagnostic tag is currently supported via LSP.
-                Tags = diagnostic.CustomTags.Contains("Unnecessary") ? new DiagnosticTag[] { DiagnosticTag.Unnecessary } : Array.Empty<DiagnosticTag>()
-            }).ToArray();
+                var filePath = diagnosticData.DataLocation?.MappedFilePath ?? document.FilePath;
```

Comment:
i hate this too.  1. Document.FilePath was never supposed to be used.    2. when does a diagnostic not have a data location??

Reply:
Not sure - but it's nullable.  Should I rather always use the original file path from the data location and just throw if null?

Reply:
> when does a diagnostic not have a data location??

I could imagine diagnostics generally may not have a location: consider things like when the compiler hits an IO error and has to report that somehow. That said, I'd imagine all the diagnostics you get from the diagnostic service here *would* since they're tied to an ID.